internal/vault/configmap.go

```diff
@@ -0,0 +1,211 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package vault
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/hashicorp/vault-secrets-operator/internal/common"
+)
+
+type ShutDownMode int
+
+type ShutDownStatus int
+
+const (
+	ShutDownModeUnknown ShutDownMode = iota
+	ShutDownModeRevoke
+	ShutDownModeNoRevoke
+
+	ConfigMapKeyShutDownMode   = "shutDownMode"
+	ConfigMapKeyShutDownStatus = "shutDownStatus"
+
+	ShutDownStatusDone ShutDownStatus = iota
+	ShutDownStatusFailed
+	ShutDownStatusPending
+	ShutDownStatusUnknown
+)
+
+func (m ShutDownMode) String() string {
+	switch m {
+	case ShutDownModeRevoke:
+		return "revoke"
+	case ShutDownModeNoRevoke:
+		return "no-revoke"
+	default:
+		return "unknown"
+	}
+}
+
+func (s ShutDownStatus) String() string {
+	switch s {
+	case ShutDownStatusPending:
+		return "pending"
+	case ShutDownStatusFailed:
+		return "failed"
+	case ShutDownStatusDone:
+		return "done"
+	default:
+		return "unknown"
+	}
+}
+
+type OnConfigMapChange func(context.Context, client.Client, *corev1.ConfigMap)
+
+func WaitForManagerConfigMapModified(ctx context.Context, watcher watch.Interface, c client.Client, onChanges ...OnConfigMapChange) {
+	defer watcher.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case event := <-watcher.ResultChan():
```

Comment:
I feel like we should return from the for loop after this triggers because any update to our configmap while we are processing shutdown will cause us to enter this path again. What do you think?

Reply:
I agree with you. I will refactor the on change function a bit to avoid the case.